### ISSUE TITLE
Implement FoVBackgroundMaker

### ISSUE THREAD

**Issue by AtreyeeS** [gammapy/gammapy#2223]
While doing stacked analysis with HESS data, I get the feeling that it is inevitable that users would like some normalisation of the background. I propose to add an option on `MapMaker`, `normalise_background=True` which will compute a field of view normalisation using the exclusion region already passed to the maker and scale the background map accordingly, before adding it onto the stacked map. 

One question is whether an average scaling should be computed over all energy bins, or if it should be computed for each energy. The problem with the later is that it suffers from huge statistical fluctuations. So I would prefer to go with an average scaling for now.

**Comment by cdeil** [comment]
Could be added here, fine with me for now to compute a norm factor and apply it here if it's not much code and work:
https://github.com/gammapy/gammapy/blob/eb595be3715a12adfd3797ed53737491ae089d00/gammapy/cube/make.py#L302-L303

Making such additions to `MapMaker` is possible and maybe we should. Eventually this will likely turn into a mess though, and a better design where different background estimation methods can be plugged into a map maker pipeline is needed, kind of like https://scikit-learn.org/stable/modules/compose.html . We had something similar a year ago, developed by @adonath with the IACTBasicImageEstimator. Maybe the ideas there were good and we just weren't set up properly yet to make it work in a clean way - e.g. ObservationFilter on the input observations was missing, and `MapDataSet` as an output object was missing. If we can achieve something where different background estimation methods can be hooked into a map data reduction pipeline, that would be better than adding code to the existing methods, because as you say: some people will want a scalar norm factor, others will want to correct also for spectral and spatial variations, ... there will be many background models for IACT data that people want to use.

@registerrier and @adonath - Do you want to spend time on such additions now? Or work on a PIG first to get to a good and agreed solution first?

**Comment by AtreyeeS** [comment]
As discussed in the dev call, the way forward is

1. An immediate PR with simple FoV normalisation at present.
2. A systematic method of background normalisation needs to be decided upon during the v0.13 cycle.

Keeping this issue open for now.

**Comment by adonath** [comment]
Unfortunately I couldn't join the dev call today. Here are a some thought from my side:

I think data reduction will be still a topic for v1.0. It's almost clear to me, that the `MapMaker` won't survive until v1.0, but will be replaced by something like a `MapDatasetEstimator` and maybe a helper method to stack multiple datasets, or similar. I think it's also  likely that stacking becomes a 2-step procedure i.e. first the datasets per observation are computed and written to disk. In a second step the 
datasets can be grouped and stacked (either by pasting or reprojecting...). This offers the possibility to e.g. pre-fit background models per observation and then stack. So I would not spend much time on this for now, but rather start working on the PIG with an API proposal for data reduction. Ideally we have a draft for this for the coding sprint Erlangen, where we can discuss it. @AtreyeeS If  you don't need this feature urgently, an immediate PR is probably not necessary...

I think in general we should handle the background correction by background model parameters and not by scaling the data. I think the scaling factors (norm / tilt) for each observation are useful information to debug / improve background models, so it should be available somehow to make distributions and plots.

**Comment by AtreyeeS** [comment]
Thanks @adonath. I think this is a decent plan 
> I think data reduction will be still a topic for v1.0. It's almost clear to me, that the `MapMaker` won't survive until v1.0, but will be replaced by something like a `MapDatasetEstimator` and maybe a helper method to stack multiple datasets, or similar. I think it's also likely that stacking becomes a 2-step procedure i.e. first the datasets per observation are computed and written to disk. In a second step the datasets can be grouped and stacked (either by pasting or reprojecting...). This offers the possibility to e.g. pre-fit background models per observation and then stack.

This sounds good. This allows a lot of flexibility in the normalisation procedure then. Also, helps keep track the norm for each run.

> @AtreyeeS If you don't need this feature urgently, an immediate PR is probably not necessary...

I have a workaround in my own scripts for now, so it is not super urgent for me personally. But this is definitely something that would be needed for users.

> I think in general we should handle the background correction by background model parameters and not by scaling the data. I think the scaling factors (norm / tilt) for each observation are useful information to debug / improve background models, so it should be available somehow to make distributions and plots.

Do you propose to make joint fitting mandatory for users to do a background scaling? On my machine, joint fits take ~ few hours, and I am not sure it is really necessary for all science cases. Simple scalings can work as well for stacked analysis in many cases...

**Comment by adonath** [comment]
@AtreyeeS No, I didn't mean to make joint fitting mandatory. The idea was rather, that once you have the individual datasets per observations you could e.g. pre-fit the background models per observation using a rough (maybe catalog-based) source model, fit the background outside exclusion regions (e.g. if there aren't many counts) or just scale it outside exclusion regions, as you proposed. This way you get a good estimate for the background model parameters for the likelihood fit, that you an also use for stacking in the 2nd step. I hope this makes sense...

**Comment by adonath** [comment]
In the new data reduction scheme this becomes possible. Here is a minimal example:

```python3
from gammapy.cube import MapDatasetMaker, MapDataset
from gammapy.modeling import Fit

maker = MapDatasetMaker(...)
stacked = MapDataset.create(...)

for obs in observations:
    dataset = maker.run(obs)
    dataset.background_model.tilt.frozen = False
    dataset.mask_fit = ...
    fit = Fit(dataset)
    fit.optimize()
    stacked.stack(dataset)
```

I think the remaining action items here are to extend the tutorials to show this example and possibly to support this functionality in the `Analysis` class...

**Comment by registerrier** [comment]
Why not create a `FoVBackgroundMaker` that would perform this? This would allow to use it from the HLI interface directly.
I guess it is not much more effort, than adapting the notebooks. No?

**Comment by adonath** [comment]
@registerrier I agree, we could implement a `FoVBackgroundMaker` or `BackgroundNormMaker` for this purpose, might be easier to integrate in the `Analysis` class and it could also offer the options `fit` and `scale` if needed...

**Comment by adonath** [comment]
I think this is an important feature for v1.0 and takes less then half a day to implement. I'll re-label the issue now...

**Comment by registerrier** [comment]
This is now implemented as the `FoVBackgroundMaker` in PR #2697 . Closing issue now.
